### DISCUSSION TITLE
How does iperf3 calculate udp throughput

### DISCUSSION THREAD

**Discussion by cyanideangel2000** [esnet/iperf#1671] [General]
Hello Team,

Quick query, is the iperf3 UDP throughput a reliable metric?
I am using iperf3 version 3.16

I ran the following command to pump packets over UDP between two hosts
./iperf3 -c 10.136.57.72 -p 2291 -t 60 -i 5 -P 8 -Z -u -b 30G

When I see the output i notice:
Connecting to host 10.136.57.72, port 2291                                                                                                                                                                          
[  5] local 10.136.16.114 port 51916 connected to 10.136.57.72 port 2291                                                                                                                                            
[  7] local 10.136.16.114 port 49101 connected to 10.136.57.72 port 2291                                                                                                                                            
[  9] local 10.136.16.114 port 40756 connected to 10.136.57.72 port 2291                                                                                                                                            
[ 11] local 10.136.16.114 port 37425 connected to 10.136.57.72 port 2291                                                                                                                                            
[ 13] local 10.136.16.114 port 53046 connected to 10.136.57.72 port 2291                                                                                                                                            
[ 15] local 10.136.16.114 port 48840 connected to 10.136.57.72 port 2291                                                                                                                                            
[ 17] local 10.136.16.114 port 35327 connected to 10.136.57.72 port 2291                                                                                                                                            
[ 19] local 10.136.16.114 port 39682 connected to 10.136.57.72 port 2291                                                                                                                                            
[ ID] Interval           Transfer     Bitrate         Total Datagrams                                                                                                                                               
[  5]   0.00-5.00   sec  1.54 GBytes  2.65 Gbits/sec  1142085                                                                                                                                                       
[  7]   0.00-5.00   sec  1.18 GBytes  2.03 Gbits/sec  875029                                                                                                                                                        
[  9]   0.00-5.00   sec  1.60 GBytes  2.75 Gbits/sec  1186688                                                                                                                                                       
[ 11]   0.00-5.00   sec  1.54 GBytes  2.64 Gbits/sec  1140354                                                                                                                                                       
[ 13]   0.00-5.00   sec  1.46 GBytes  2.52 Gbits/sec  1086011                                                                                                                                                       
[ 15]   0.00-5.00   sec  1.07 GBytes  1.84 Gbits/sec  793978                                                                                                                                                        
[ 17]   0.00-5.00   sec  1.39 GBytes  2.38 Gbits/sec  1028043                                                                                                                                                       
[ 19]   0.00-5.00   sec  1.36 GBytes  2.34 Gbits/sec  1008826                                                                                                                                                       
[SUM]   0.00-5.00   sec  11.1 GBytes  19.1 Gbits/sec  8261014                                                                                                                                                       
- - - - - - - - - - - - - - - - - - - - - - - - -                                                                                                                                                                   
[  5]   5.00-10.00  sec  1.52 GBytes  2.62 Gbits/sec  1129406                                                                                                                                                       
[  7]   5.00-10.00  sec  1.16 GBytes  2.00 Gbits/sec  862803                                                                                                                                                        
[  9]   5.00-10.00  sec  1.79 GBytes  3.08 Gbits/sec  1330759                                                                                                                                                       
[ 11]   5.00-10.00  sec  1.52 GBytes  2.61 Gbits/sec  1126821                                                                                                                                                       
[ 13]   5.00-10.00  sec  1.49 GBytes  2.55 Gbits/sec  1102738                                                                                                                                                       
[ 15]   5.00-10.00  sec  1.03 GBytes  1.76 Gbits/sec  760682                                                                                                                                                        
[ 17]   5.00-10.00  sec  1.37 GBytes  2.35 Gbits/sec  1014828                                                                                                                                                       
[ 19]   5.00-10.00  sec  1.36 GBytes  2.33 Gbits/sec  1006596                                                                                                                                                       
[SUM]   5.00-10.00  sec  11.2 GBytes  19.3 Gbits/sec  8334633 

=====
[SUM]   0.00-60.00  sec   135 GBytes  19.4 Gbits/sec  0.000 ms  0/100291189 (0%)  sender                  
[SUM]   0.00-60.00  sec  98.2 GBytes  14.1 Gbits/sec  0.007 ms  27446878/100288639 (27%)  receiver               

Mine is a single connection upload test, however I see throughput being reported for both sender and receiver.
How do we read the throughput numbers for sender and receiver.
Does it mean the sender could pump at the rate of 19.4Gbps out of the node, however receiver due to 27% loss was receiving at the rate of 14Gbps?

Thank you.

**Comment by davidBar-On** [comment]
Hi @cyanideangel2000,
> Does it mean the sender could pump at the rate of 19.4Gbps out of the node, however receiver due to 27% loss was receiving at the rate of 14Gbps?

In general this is the case.  The client shows the sent throughput and the server shows the received throughput, which s affected by the network throughput.

If you want to better understand the performance you may try using different packet sizes (`-l` option).  For UDP the default is usually 1460 bytes.  Larger sizes (e.g. 8K, 32K) may improve throughput, although this is not for sure because of the added fragmentation.  (Note by the way that `-Z` is applicable only for TCP).

Also, set `-b` to just above the expected throughput per stream.  E.g. `-b 3G` per the results you sent.  This is because trying to send redundant messages overload the CPU / buffers / etc. and reduces performance.

Another thing to try is changing the number of streams (`-P`), as if several streams run on the same processor, that may result in lower throughput.
